### PR TITLE
Ppo

### PR DESCRIPTION
**Pull request by nelsonayamashita**
Coloquei o notebook com a implementação de PPO da aula, mas acredito que ela não esteja funcionando 100% (vide gráfico no fim do nb).

### CODE REVIEW

**Comment by Berbardo** [comment]
Por que essa branch está com todas as mudanças da branch de DQN?

**Comment by fernandokm** [comment]
> Por que essa branch está com todas as mudanças da branch de DQN?

Corrigido.

File: Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+    "        old_logp = torch.FloatTensor(old_logp).to(self.device)\n",
+    "\n",
+    "        for epoch in range(self.epochs):\n",
+    "            probs, v = self.actorcritic.forward(states)\n",
+    "            with torch.no_grad():\n",
+    "                _, v2 = self.actorcritic.forward(next_states)\n",
+    "\n",
+    "            new_logp = probs.log_prob(actions)\n",
+    "\n",
+    "            advantages, returns = self.compute_gae(rewards, dones, v, v2)\n",
```
Comment (Berbardo) [inline Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb:305]: Aqui a gente calcula o v2 e o GAE a cada epoch mesmo? Ou só uma vez antes do loop?

As implementações que eu costumo ver calculam antes, mas talvez calcular cada vez faça sentido considerando que vamos ter um v e um v2 mais atualizados.

File: Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+    "    avg_returns = []\n",
+    "\n",
+    "    state = env.reset()\n",
+    "    timestep = 0\n",
+    "    episode = 0\n",
+    "\n",
+    "    while timestep < total_timesteps:\n",
+    "        action, log_prob = agent.act(state)\n",
+    "        next_state, reward, done, _ = env.step(action)\n",
+    "        agent.remember(state, action, reward, next_state, done, log_prob.detach().cpu().numpy())\n",
```
Comment (Berbardo) [inline Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb:352]: Acho que esse .detach().cpu().numpy() fica meio jogado aqui, talvez seja melhor colocar ele no act

File: Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+    "            Estado da transição.\n",
+    "        action: int\n",
+    "            Ação tomada.\n",
+    "        reward: float\n",
+    "            Recompensa recebida.\n",
+    "        state: np.array\n",
+    "            Estado seguinte.\n",
+    "        done: int\n",
+    "            Flag indicando se o episódio acabou.\n",
+    "        \"\"\"\n",
```
Comment (Berbardo) [inline Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb:194]: As docstrings do Replay não estão com o ```logp```

File: Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+    "    \\quad\n",
+    "    g(\\varepsilon, A) = \\begin{cases}\n",
+    "        (1+\\varepsilon) A, & A \\ge 0 \\\\\n",
+    "        (1-\\varepsilon) A,  & A < 0.\n",
+    "    \\end{cases}\n",
+    "$$\n",
+    "\n",
+    "Nota-se que:\n",
+    "- Quando a vantagem é positiva, se $r$ aumentar, então $L$ aumenta. No entanto, esse benefício é limitado pelo clip: se $r > 1+\\varepsilon$, não há mais benefício para $r$ aumentar.\n",
+    "- Quando a vantagem é positiva, se $r$ diminuir, então $L$ aumenta. No entanto, esse benefício é limitado pelo clip: se $r M 1-\\varepsilon$, não há mais benefício para $r$ diminuir."
```
Comment (Berbardo) [inline Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb:99]: Aqui era pra ser $r M 1-\\varepsilon$ mesmo?

**Review by Berbardo — CHANGES_REQUESTED** [review]

File: Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb (hunk 4 above)
Comment (Berbardo) [reply Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb:99]: E acho que talvez devesse ser "Quando a vantagem é *negativa*, se $r$ diminuir..."

Comment (Berbardo) [reply Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb:99]: Talvez usar uma imagenzinha pra explicar tipo essa:

**Comment by fernandokm** [comment]
Consequências de colocar direto o notebook que a gente usou na aula sem terminar de arrumar as coisas... Apaguei os arquivos do vscode, depois eu arrumo o resto.

File: Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+    "        rewards = torch.FloatTensor(rewards).unsqueeze(-1).to(self.device)\n",
+    "        next_states = torch.FloatTensor(next_states).to(self.device)\n",
+    "        dones = torch.FloatTensor(dones).unsqueeze(-1).to(self.device)\n",
+    "        old_logp = torch.FloatTensor(old_logp).to(self.device)\n",
+    "        \n",
+    "        with torch.no_grad():\n",
+    "            _, v = self.actorcritic.forward(states)\n",
+    "            _, v2 = self.actorcritic.forward(next_states)\n",
+    "        \n",
+    "        advantages, returns = self.compute_gae(rewards, dones, v, v2)\n",
```
Comment (nelsonayamashita) [inline Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb:306]: Existe um jeito mais elegante e mais eficiente de pegar v e v2 sem ter que dar esse forward? Achei que ficou meio desnecessário mas foi o que eu pensei na hora.

Comment (Berbardo) [reply Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb:310]: Como os estados estão em ordem cronológica, o v2[i] é sempre o v[i+1]. Se você quiser otimizar ao máximo dá pra colocar o último v2 no v e passar tudo de uma vez no ```forward```, aí mudar o gae pra usar só o v[i] e o v[i+1]. Mas não sei se tem muita coisa pra melhorar ;P, esse que eu comentei já pode ficar um pouco estranho também.

Comment (fernandokm) [reply Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb:310]: Não precisaria necessariamente mudar o gae. Da pra fazer isso:
```python
with torch.nograd():
    _, vs = self.actorcritic.forward(numpy.append(states, next_states[[-1]]))
v1 = vs[:-1]
v2 = vs[1:]
```

**Comment by nelsonayamashita** [comment]
Coisas que podemos melhorar depois da aula de PPO:
 - [ ] Trocar os nomes referentes a on-policy (tirar ReplayBuffer e Sample)
 - [ ] Trocar tangente hiperbólica para ReLu na rede de valor do estado
   - [ ] Rodar o agente com essas mudanças

**Comment by nelsonayamashita** [comment]
Troquei os nomes e rodei o a gente. Eu suspeito que o gráfico tava desatualizado, quando eu rodei com as camadas relus o resultado foi horrível. Adaptei o parâmetro do tamanho da memória e ele pareceu melhorar. Talvez dê para melhorar o modelo com alguns hiper-parâmetros, mas sem tempo, irmão

**Comment by nelsonayamashita** [comment]
Eu fiz merda dando merge da master aqui na branch, mas pelo menos com isso deu para adicionar o texto de PPO e colocar ele na pasta certo junto com o A2C. Aproveitei e tbm arrumei o nome das classes do A2C que tbm tavam como "ExperienceReplay" e "sample".

File: Aprendizado por Reforço Profundo/Actor-Critic/PPO/README.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+<img src="https://latex.codecogs.com/svg.latex?\quad&space;g(\varepsilon,&space;A)&space;=&space;\begin{cases}&space;(1&plus;\varepsilon)&space;A,&space;&&space;A&space;\ge&space;0&space;\\&space;(1-\varepsilon)&space;A,&space;&&space;A&space;<&space;0.&space;\end{cases}" title="\quad g(\varepsilon, A) = \begin{cases} (1+\varepsilon) A, & A \ge 0 \\ (1-\varepsilon) A, & A < 0. \end{cases}" />
+
+
+Nota-se que:
+- Quando a vantagem é positiva, se **r** aumentar, então **L** aumenta. No entanto, esse benefício é limitado pelo clip: se **r > 1+&epsilon;**, não há mais benefício para **r** aumentar.
+- Quando a vantagem é negativa, se **r** diminuir, então **L** aumenta. No entanto, esse benefício é limitado pelo clip: se **r < 1-&epsilon;**, não há mais benefício para **r** diminuir.
+
+A seguinte imagem pode te ajudar a visualizar o clip. Note que todos os valores fora do clip estipulado estão constantes:
+
+![imagem ilustrando o clip](imgs/clip.png)
```
Comment (Berbardo) [inline Aprendizado por Reforço Profundo/Actor-Critic/PPO/README.md:62]: Está sem o algoritmo de PPO em LaTeX (não sei se precisa)

File: Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+    "\t\t&= E_{\\pi_{\\theta_{\\mathrm{old}}}} \\left[\\frac{\\pi_\\theta(a|s)}{\\pi_{\\theta_{\\mathrm{old}}}(a|s)} A^{\\pi_\\theta}(s,a)\\right]\n",
+    "\\end{align*}\n",
+    "\n",
+    "\n",
+    "## Proximal Policy Optimization (PPO)\n",
+    "\n",
+    "Como já foi mencionado, a restrição ($KL < \\delta$) imposta em TRPO torna o algoritmo relativamente complicado. PPO é uma tentativa de simplificar esse algoritmo. Ao invés de utilizar trust regions, PPO mexe diretamente com a função objetivo:\n",
+    "\n",
+    "$$\n",
+    "    L(\\theta_{\\mathrm{old}},\\theta) = E_{s,a\\sim\\pi_{\\theta_{\\mathrm{old}}}} \\Bigl[\\min\\left(r A^{\\pi_{\\theta_{\\mathrm{old}}}}(s,a),\\, \\operatorname{clip}(r,1-\\varepsilon,1+\\varepsilon) A^{\\pi_{\\theta_{\\mathrm{old}}}}(s,a)\\right)\\Bigr],\n",
```
Comment (Berbardo) [inline Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb:78]: não entendi esse (r, 1-eps, 1+eps)

File: Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+    "    \\end{cases}\n",
+    "$$\n",
+    "\n",
+    "Nota-se que:\n",
+    "- Quando a vantagem é positiva, se $r$ aumentar, então $L$ aumenta. No entanto, esse benefício é limitado pelo clip: se $r > 1+\\varepsilon$, não há mais benefício para $r$ aumentar.\n",
+    "- Quando a vantagem é negativa, se $r$ diminuir, então $L$ aumenta. No entanto, esse benefício é limitado pelo clip: se $r < 1-\\varepsilon$, não há mais benefício para $r$ diminuir.\n",
+    "\n",
+    "A seguinte imagem pode te ajudar a visualizar o clip. Note que todos os valores fora do clip estipulado estão constantes:\n",
+    "\n",
+    "![imagem ilustrando o clip](imgs/clip.png)\n"
```
Comment (Berbardo) [inline Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb:98]: Acho que compensa explicar melhor o resultado prático disso

File: Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import torch.nn as nn\n",
+    "import torch.nn.functional as F\n",
+    "from torch.distributions import Categorical\n",
+    "class ActorCritic(nn.Module):\n",
```
Comment (fernandokm) [inline Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb:117]:
```suggestion
    "from torch.distributions import Categorical\n",
    "\n",
    "class ActorCritic(nn.Module):\n",
```

File: Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+    "        action: int\n",
+    "            Ação tomada.\n",
+    "        reward: float\n",
+    "            Recompensa recebida.\n",
+    "        state: np.array\n",
+    "            Estado seguinte.\n",
+    "        done: int\n",
+    "            Flag indicando se o episódio acabou.\n",
+    "        logp: float\n",
+    "            Log da probabildiade de acordo com a política.\n",
```
Comment (fernandokm) [inline Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb:194]:
```suggestion
    "            Log da probabilidade de acordo com a política.\n",
```

File: Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+    "        actions: np.array\n",
+    "            Batch de ações.\n",
+    "        rewards: np.array\n",
+    "            Batch de recompensas.\n",
+    "        next_states: np.array\n",
+    "            Batch de estados seguintes.\n",
+    "        dones: np.array\n",
+    "            Batch de flags indicando se o episódio acabou.\n",
+    "        logp: np.array\n",
+    "            Batch do log da probabildiade de acordo com a política.\n",
```
Comment (fernandokm) [inline Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb:225]:
```suggestion
    "            Batch do log da probabilidade de acordo com a política.\n",
```

File: Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb [hunk 12]
```diff
@@ -0,0 +1,440 @@
...
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import torch\n",
+    "import torch.optim as optim\n",
+    "\n",
+    "class PPO:\n",
+    "    def __init__(self, observation_space, action_space, lr=7e-4, gamma=0.99, lam=0.95, vf_coef=0.5, entropy_coef=0.005,clip_param =0.2, epochs =10, memory_len=16):\n",
```
Comment (fernandokm) [inline Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb:242]:
```suggestion
    "    def __init__(self, observation_space, action_space, lr=7e-4, gamma=0.99, lam=0.95,\n",
    "                 vf_coef=0.5, entropy_coef=0.005, clip_param=0.2, epochs=10, memory_len=16):\n",
```

**Review by fernandokm — CHANGES_REQUESTED** [review]

File: Aprendizado por Reforço Profundo/Actor-Critic/PPO/README.md (hunk 6 above)
Comment (fernandokm) [reply Aprendizado por Reforço Profundo/Actor-Critic/PPO/README.md:62]: Acho legal por consistência, mas não vejo muito problema em não ter

File: Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+    "        dists = torch.relu(self.policy1(state))\n",
+    "        dists = torch.relu(self.policy2(dists))\n",
+    "        dists = F.softmax(self.policy3(dists), dim=-1)\n",
+    "        probs = Categorical(dists)\n",
+    "        \n",
+    "        v = torch.relu(self.value1(state))\n",
+    "        v = torch.relu(self.value2(v))\n",
+    "        v = self.value3(v)\n",
+    "\n",
+    "        return probs, v"
```
Comment (FelipeAugustoMachado) [inline Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb:138]: Adicionar comentários, mesmo que pequenos, para ficar mais explicativo. O do Buffer possui tantos e as outras classes não possuem nenhum, tadinha delas :(((

File: Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb [hunk 14]
```diff
@@ -0,0 +1,440 @@
...
+    "\n",
+    "            policy_loss =   - torch.min(surr1,surr2).mean()\n",
+    "            value_loss =    self.vf_coef * F.mse_loss(v, returns.detach())\n",
+    "            entropy_loss = -self.entropy_coef * entropy\n",
+    "\n",
+    "            self.actorcritic_optimizer.zero_grad()\n",
+    "            (policy_loss + entropy_loss + value_loss).backward()\n",
+    "            self.actorcritic_optimizer.step()\n",
+    "\n",
+    "        return policy_loss + entropy_loss + value_loss"
```
Comment (FelipeAugustoMachado) [inline Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb:328]: Colocar comentários também, principalmente nas linhas mais importantes (relacionadas com a teoria).

File: Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb [hunk 15]
```diff
@@ -0,0 +1,440 @@
...
+    "\n",
+    "plt.plot(returns, 'r')\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "O agente demonstra consegue solocuinar o ambiente. Porém acaba desaprendendo, para solucionar isso é possível fazer uma otimização nos hiper-parâmetros. É possível também implementar algum tipo de parada antecipada."
```
Comment (FelipeAugustoMachado) [inline Aprendizado por Reforço Profundo/Actor-Critic/PPO/PPO.ipynb:436]:
```suggestion
    "O agente demonstra consegue solocuinar o ambiente. Porém acaba desaprendendo, para solucionar isso é possível fazer uma otimização nos hiper-parâmetros. É possível também implementar algum tipo de parada antecipada."
```
"O agente aparentemente consegue solucinar o ambiente, no entanto, acaba desaprendendo. Para solucionar isso, é possível fazer uma otimização nos hiper-parâmetros ou implementar algum tipo de parada antecipada."

**Review by FelipeAugustoMachado — CHANGES_REQUESTED** [review]

File: Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb (hunk 5 above)
Comment (Berbardo) [reply Aprendizado por Reforço Profundo/Actor Critic/PPO/PPO.ipynb]: @nelsonayamashita @fernandokm mexemos nisso aqui?
